src/ModelContextProtocol/Auth/OAuthDelegatingHandler.cs

```diff
@@ -0,0 +1,140 @@
+    /// <param name="scopes">The requested scopes.</param>
+    /// <param name="authorizationHandler">A handler to invoke when authorization is required.</param>
+    public OAuthDelegatingHandler(
+        Uri redirectUri,
+        string? clientId = null,
+        string? clientName = null,
+        IEnumerable<string>? scopes = null,
+        Func<Uri, Task<string>>? authorizationHandler = null)
+    {
+        _redirectUri = redirectUri;
+        _clientId = clientId;
+        _clientName = clientName;
+        _scopes = scopes;
+        _authorizationHandler = authorizationHandler;
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="OAuthDelegatingHandler"/> class with an inner handler.
+    /// </summary>
+    /// <param name="innerHandler">The inner handler which processes the HTTP response messages.</param>
+    /// <param name="redirectUri">The URI to redirect to after authentication.</param>
+    /// <param name="clientId">The client ID to use for authentication, or null to register a new client.</param>
+    /// <param name="clientName">The client name to use for registration.</param>
+    /// <param name="scopes">The requested scopes.</param>
+    /// <param name="authorizationHandler">A handler to invoke when authorization is required.</param>
+    public OAuthDelegatingHandler(
+        HttpMessageHandler innerHandler,
+        Uri redirectUri,
+        string? clientId = null,
+        string? clientName = null,
+        IEnumerable<string>? scopes = null,
+        Func<Uri, Task<string>>? authorizationHandler = null)
+        : base(innerHandler)
+    {
+        _redirectUri = redirectUri;
+        _clientId = clientId;
+        _clientName = clientName;
+        _scopes = scopes;
+        _authorizationHandler = authorizationHandler;
+    }
+
+    /// <summary>
+    /// Manually set an OAuth token to use for subsequent requests.
+    /// </summary>
+    /// <param name="tokenResponse">The OAuth token response.</param>
+    public void SetToken(OAuthToken tokenResponse)
+    {
+        _tokenResponse = tokenResponse;
+    }
+
+    /// <inheritdoc />
+    protected override async Task<HttpResponseMessage> SendAsync(
+        HttpRequestMessage request,
+        CancellationToken cancellationToken)
+    {
+        // If we have a token, attach it to the request
+        if (_tokenResponse != null)
+        {
+            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _tokenResponse.AccessToken);
+        }
+
+        // Send the request
+        var response = await base.SendAsync(request, cancellationToken);
+
+        // If the response is 401 Unauthorized, try to authenticate
+        if (response.StatusCode == HttpStatusCode.Unauthorized)
+        {
+            // Create a temporary HttpClient to handle the authentication
+            // We need to use a new client to avoid infinite recursion
+            using var httpClient = new HttpClient();
```

Comment:
a new http client should not be created within a delegating handler. Use base.SendAsync to make other network calls.